crates/sui-types/src/gas.rs

```diff
@@ -234,14 +249,20 @@ impl<'a> SuiGasStatus<'a> {
     /// computation cost.
     pub fn summary(&self, succeeded: bool) -> GasCostSummary {
         let remaining_gas = self.gas_status.remaining_gas();
-        let storage_cost = self.storage_cost;
-        let computation_cost = self.init_budget.sub(remaining_gas).sub(storage_cost);
-        let computation_cost_in_sui = computation_cost.mul(self.computation_gas_unit_price).get();
+        let storage_cost = self.storage_gas_units;
+        // TODO: handle overflow how?
+        let computation_cost = self
+            .init_budget
+            .checked_sub(remaining_gas)
+            .expect("Subtraction overflowed")
```

Comment:
Should we really be panicking here?

Reply:
Right. What should we do?

Reply:
Looks like outside of tests, this function is only used in `sui_core::execution_engine::execute_transaction` which returns `ExecutionError` as part of the result.

Perhaps in case of overflow, we can return the "failed summary" and combine it with propagating `ExecutionError` to `execute_transaction` and beyond.

Reply:
@awelc but what will be in the summary? Do we count the underflow/overflow result?

Reply:
> @awelc but what will be in the summary? Do we count the underflow/overflow result? Btw this was a problem all along. We just werent handling it. I will flesh out a better soln to this in[ next PR](https://github.com/MystenLabs/sui/issues/4172)

Sounds good. And I don't fully understand the logic there but I thought that we can treat underflow/overflow as execution failure and return whatever we return when `succeeded` is false, that is:
```
            GasCostSummary {
                computation_cost: computation_cost_in_sui,
                storage_cost: 0,
                storage_rebate: 0,
            }
```

Reply:
An underflow shouldn't happen here. This is already at the end of execution, and we were able to go this far because the gas charged did not exceed the budget.

Reply:
The gas algebra logic removed the ability to do an `unchecked_sub`, so we can only call `checked_sub` and since guaranteed to not underflow, we can unwrap indeed.